proof/invariant-abstract/DetSchedSchedule_AI.thy

```diff
@@ -7177,11 +7209,15 @@ lemma refill_head_overlapping_loop_hd_r_amount:
   done
 
 lemma merge_refills_no_overflow_helper:
-  "\<lbrace>\<lambda>s. pred_map (\<lambda>cfg. unat (r_time (scrc_refill_hd cfg)) + refills_unat_sum (scrc_refills cfg)
-                        \<le> unat max_time)
-                 (sc_refill_cfgs_of s) sc_ptr
-        \<and> pred_map (\<lambda>cfg. Suc 0 < length (scrc_refills cfg)) (sc_refill_cfgs_of s) sc_ptr\<rbrace>
-   merge_refills sc_ptr
+  "\<lbrace>\<lambda>s. if sc_ptr' = sc_ptr
+        then pred_map (\<lambda>cfg. unat (r_time (scrc_refill_hd cfg)) + refills_unat_sum (scrc_refills cfg)
+                             \<le> unat max_time)
+                      (sc_refill_cfgs_of s) sc_ptr
+             \<and> pred_map (\<lambda>cfg. Suc 0 < length (scrc_refills cfg)) (sc_refill_cfgs_of s) sc_ptr
+        else pred_map (\<lambda>cfg. unat (r_time (scrc_refill_hd cfg)) + refills_unat_sum (scrc_refills cfg)
+                          \<le> unat max_time)
+                   (sc_refill_cfgs_of s) sc_ptr\<rbrace>
+   merge_refills sc_ptr'
```

Comment:
The same here. I will stop pointing out the same issues for the rest of the PR, but please refactor all the `if P then A & B else A` instances in this PR to `A & (P --> B)`. These `A`s are often quite big and repeating them affects readability, and it is always better to be clear about what is common and what is not. Also, when we need to make modifications in the future, we don’t want to be spending time on duplicating changes or going through all of them to see which was is the same or not.

Reply:
I've pushed a commit which rearranges the preconditions, but I'm still curious about this idea with `if_helper`. I think there are many situations where we could use this approach